homeassistant/components/sensor/airvisual.py

```diff
@@ -173,127 +206,88 @@ def state(self):
+                ATTR_POLLUTANT_UNIT: POLLUTANT_MAPPING[symbol]['unit']
+            })
 
 
 class AirVisualData(object):
     """Define an object to hold sensor data."""
 
     def __init__(self, client, **kwargs):
-        """Initialize the AirVisual data element."""
+        """Initialize."""
         self._client = client
-        self.attrs = {}
-        self.pollution_info = None
-
         self.city = kwargs.get(CONF_CITY)
-        self.state = kwargs.get(CONF_STATE)
         self.country = kwargs.get(CONF_COUNTRY)
-
         self.latitude = kwargs.get(CONF_LATITUDE)
         self.longitude = kwargs.get(CONF_LONGITUDE)
-        self._radius = kwargs.get(CONF_RADIUS)
-
+        self.pollution_info = {}
         self.show_on_map = kwargs.get(CONF_SHOW_ON_MAP)
+        self.state = kwargs.get(CONF_STATE)
+
+        self.async_update = Throttle(
+            kwargs[CONF_SCAN_INTERVAL])(self._async_update)
 
-    @Throttle(MIN_TIME_BETWEEN_UPDATES)
-    def update(self):
-        """Update with new AirVisual data."""
-        from pyairvisual.exceptions import HTTPError
+    async def _async_update(self):
+        """Update AirVisual data."""
+        from pyairvisual.errors import AirVisualError
 
         try:
             if self.city and self.state and self.country:
-                resp = self._client.city(
-                    self.city, self.state, self.country).get('data')
-                self.longitude, self.latitude = resp.get('location').get(
-                    'coordinates')
+                resp = await self._client.data.city(
+                    self.city, self.state, self.country)
+                self.longitude, self.latitude = resp['location']['coordinates']
             else:
-                resp = self._client.nearest_city(
-                    self.latitude, self.longitude, self._radius).get('data')
+                resp = await self._client.data.nearest_city(
+                    self.latitude, self.longitude)
+
             _LOGGER.debug("New data retrieved: %s", resp)
 
-            self.pollution_info = resp.get('current', {}).get('pollution', {})
-
-            self.attrs = {
-                ATTR_CITY: resp.get('city'),
-                ATTR_REGION: resp.get('state'),
-                ATTR_COUNTRY: resp.get('country')
-            }
-        except HTTPError as exc_info:
-            _LOGGER.error("Unable to retrieve data on this location: %s",
-                          self.__dict__)
-            _LOGGER.debug(exc_info)
+            self.pollution_info = resp['current']['pollution']
+        except AirVisualError as err:
+            if self.city and self.state and self.country:
+                location = (self.city, self.state, self.country)
+            else:
+                location = (self.latitude, self.longitude)
+
+            _LOGGER.error(
+                "Can't retrieve data for location: %s (%s)", location,
+                err)
+            self.city = None
```

Comment:
Why would you reset all these things ? If something is misconfigured, we should not just try to get the nearest city. We should just not update and mark as available=False.

Reply:
Great point; not sure what I was thinking there.

Reply:
Let me know if you're okay with the method I use in the latest commit.